### ISSUE TITLE
[Hold for Payment June 19][$250] Workspace -  Inconsistency in Block Quote Usage Between Workspace Description and Chat

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#40571]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.63-7
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** abebemiherat@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Navigate to profile > workspace
2. Go to Description, erase the default statement, and add a block quote markdown (e.g., ">>abebe"). Notice it displays correctly
3. Click on Description again, and notice that one quote is reduced
4. Go to any chat, write the same block quote as in the description, click on edit, and then save it. Notice that it doesn't change

## Expected Result:
The block quote should be displayed correctly in the workspace description section, like in chat, ensuring consistency

## Actual Result:
When writing a block code example (">>abebe") in the workspace description, clicking on it and saving it without changes reduces one quote. However, this inconsistency does not occur in chat

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/201d7d6a-3141-40be-8cd5-9de8cbf3504a

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015e52952078f428b3</li>
        <li>Upwork Job ID: 1781372065227497472</li>
        <li>Last Price Increase: 2024-04-19</li>
<li>Automatic offers: </li>
<ul>
<li>jjcoffee | Reviewer | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @jjcoffee</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp

**Comment by skyweb331** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Inconsistency in Block Quote Usage in text
### What is the root cause of that problem?
https://github.com/Expensify/App/blob/c1827d8b169f3187a53b4f5bcc980cfb5dd1bd97/src/pages/workspace/WorkspaceProfileDescriptionPage.tsx#L29-L40

`parser.htmlToMarkdown` does not parse html correctly.

For example
`parser.htmlToMarkdown('<blockquote><blockquote>abc</blockquote></blockquote>')` => `>abc` which should be `>>abc`

### What changes do you think we should make in order to solve the problem?
Need to fix `ExpensiMark` module.

FYI, this problem also exists on chat, but if message is saved without changes, it does not call API. That's why this problem does not detected by @lanitochka17. If user edits a blockquote message on chat, it eats blockquotes too.
### What alternative solutions did you explore? (Optional)
N/A

**Comment by saifelance** [comment]
@skyweb331 

The root cause of the problem is that the parser.htmlToMarkdown function does not correctly parse HTML. Specifically, when encountering nested <blockquote> elements, it produces incorrect Markdown output.

To solve this issue, you'll need to fix the ExpensiMark module, which is responsible for parsing HTML into Markdown. The problematic behavior is identified in the ExpensiMark file, particularly in the section that handles blockquote elements.

You can address the issue by modifying the logic in the ExpensiMark.js file to correctly handle nested blockquote elements and ensure that the Markdown output reflects the correct nesting level.

**Comment by skyweb331** [comment]
@jjcoffee I am going to use `fast-xml-parser` or `xml2js`  to parse HTML string to javascript object and then will make markdown string from that object.
Regex is really difficult to handle complicated html structures.

**Comment by skyweb331** [comment]
If these modules are too big, I can just copy ideas from those modules for parsing html.

**Comment by skyweb331** [comment]
@dragnoir I just tested your code against all testing cases on expensify-common  but it fails...
your parse logic does not converts all cases correctly.

For example, `<blockquote>123<br/><blockquote>abc</blockquote></blockquote>` becomes `>>123\n>abc` by your logic which is wrong.

**Comment by skyweb331** [comment]
`>123\n>>abc` is what I input and `<blockquote>123<br/><blockquote>abc</blockquote></blockquote>` what ExpensiMark converted.

**Comment by jjcoffee** [comment]
Reviewing tomorrow!

**Comment by isabelastisser** [comment]
Hey @jjcoffee, can you please review the updated proposal above? Thanks!

**Comment by jjcoffee** [comment]
@dragnoir's [proposal](https://github.com/Expensify/App/issues/40571#issuecomment-2070373808) LGTM! Let's make sure to add extra test cases to cover this issue as well as [this one](https://github.com/Expensify/App/issues/40571#issuecomment-2070710059).

:ribbon::eyes::ribbon: C+ reviewed

**Comment by MonilBhavsar** [comment]
> The block quote should be displayed correctly in the workspace description section, like in chat, ensuring consistency

Given that parsing is working fine in chat, and not only on Workspace description, how can Expensimark parser  be affected 🤔

**Comment by jjcoffee** [comment]
> Given that parsing is working fine in chat, and not only on Workspace description, how can Expensimark parser be affected 🤔

@MonilBhavsar If you go to edit a chat message you'll see the same issue.

**Comment by MonilBhavsar** [comment]
@jjcoffee I am not seeing that issue with chats. May be I am missing something here


Uploading Screen Recording 2024-04-26 at 8.04.19 PM.mov…

**Comment by MonilBhavsar** [comment]
> The composer checks if old is equal to new then no update to the code inside the database.

Okay, thanks for clarifying. Then i think we should also fix that here and add that check in workspace description editor to also be consistent with the composer.

**Comment by MonilBhavsar** [comment]
Can you include that in your proposal please

**Comment by MonilBhavsar** [comment]
Thanks! We can work out that in PR. Should be a minor check like in composer.

**Comment by jjcoffee** [comment]
Okay, just a heads up I'll be OOO 6-13th (I may have some time on the 6th), so would be good to get a PR up today or tomorrow if you can! @dragnoir

**Comment by MonilBhavsar** [comment]
@dragnoir feel free the share the issues encountered here so we can work through it

**Comment by isabelastisser** [comment]
@jjcoffee @dragnoir is there an update? Thanks!

**Comment by MonilBhavsar** [comment]
Awaiting PR

**Comment by MonilBhavsar** [comment]
App should be working!

**Comment by MonilBhavsar** [comment]
Do you have a draft PR you're working on? If so, could you please link it

**Comment by MonilBhavsar** [comment]
Agree, the other mentioned issue looks related to live markdown.
@dragnoir feel free to post questions or concerns here

**Comment by MonilBhavsar** [comment]
@dragnoir any updates to share?

**Comment by MonilBhavsar** [comment]
Discussing in slack https://expensify.slack.com/archives/C01GTK53T8Q/p1716193746214989

**Comment by MonilBhavsar** [comment]
PR is being reviewed here

**Comment by isabelastisser** [comment]
Any updates, @dragnoir ?

**Comment by jjcoffee** [comment]
@isabelastisser App [PR](https://github.com/Expensify/App/pull/42842) is up, but we are waiting to proceed because an earlier PR to upgrade `expensify-common` recently got [reverted](https://github.com/Expensify/App/issues/42161#issuecomment-2142605342) so we don't want to upgrade until that's fixed.

**Comment by isabelastisser** [comment]
Thanks for the update, @jjcoffee !

**Comment by blazejkustra** [comment]
@jjcoffee Would you mind testing related changes to this issue on the new PR?

**Comment by MonilBhavsar** [comment]
I think it should be 7 days post https://github.com/Expensify/App/pull/42387 is deployed to production

**Comment by jjcoffee** [comment]
> Would you mind testing related changes to this issue on the new PR?

@blazejkustra I see it's already been merged, but I've given it a quick test run anyway and all looks good!

https://github.com/Expensify/App/assets/27287420/563fce11-f61a-4457-894d-1a5bc3bf7aa8

**Comment by mallenexpensify** [comment]
Can't figure out what's going on here.  @jjcoffee @MonilBhavsar can you provide an update?  Is payment due to anyone? Thx

**Comment by jjcoffee** [comment]
@mallenexpensify https://github.com/Expensify/App/pull/42387 bumped `expensify-common` to the version that includes our changes. It just hit production yesterday, so payment should be due 19 June (assuming no regressions, of course!).

**Comment by MonilBhavsar** [comment]
Waiting for payment due here.
@jjcoffee mind doing checklist please when you have a moment

**Comment by MonilBhavsar** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@jjcoffee] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@jjcoffee] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@jjcoffee] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@jjcoffee] Determine if we should create a regression test for this bug.
- [ ] [@jjcoffee] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@isabelastisser ] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by jjcoffee** [comment]
- The PR that introduced the bug has been identified. Link to the PR: **N/A - I think this behaviour has always been there**
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **N/A**
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/A**
- Determine if we should create a regression test for this bug. **Yes**
- If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Regression Test Proposal**

1. Navigate to Profile > Workspace
2. Go to Description and enter some nested block quote markdown, e.g. `>> ABC`
3. Verify that the live markdown is rendering the nested blockquotes correctly, i.e.
>> ABC
4. Save the description
5. Verify that the blockquotes still render correctly
6. Tap on Description to edit it again
7. Verify that the value to edit displays the correct number of arrows to match the nesting level of the blockquote

Do we agree 👍 or 👎

**Comment by mallenexpensify** [comment]
Contributor: @dragnoir paid $250 via Upwork
Contributor+: @jjcoffee paid $250 via Upwork

TestRail GH
- https://github.com/Expensify/Expensify/issues/405754

Thanks y'all.
